Importer/CsvImporter.php

```diff
@@ -23,88 +27,237 @@ class CsvImporter extends AbstractImporter
-     * @param bool   $hasHeaders
+     * @param string $csv
      * @param bool   $dryRun
-     *
-     * @return array|null|object[]
-     * @throws DatabaseException
      */
-    public function importFile($filename, $hasHeaders = true, $dryRun = false)
+    public function import($csv, $dryRun = false)
     {
-        $csv = file_get_contents($filename);
         try {
-            return $this->import($csv, $hasHeaders, $dryRun);
-        } catch (DatabaseException $exception) {
-            $exception->setDataFile($filename);
-            throw $exception;
+            if ($this->csvHasHeaders() && !$this->checkHeadersAreValid($csv)) {
+                return;
+            }
+
+            $this->log->setCsv($csv);
+
+            $parsedData = $this->parser->parse($csv, $this->csvHasHeaders(), $this->delimiter);
+
+            $this->importData($parsedData, $this->csvHasHeaders(), !$dryRun);
+        } catch (\Throwable $throwable) {
+            $message = sprintf(
+                '%s: %s (uncaught exception) at %s line %d while running import.',
+                get_class($throwable),
+                $throwable->getMessage(),
+                $throwable->getFile(),
+                $throwable->getLine()
+            );
+            $this->log->addCriticalError($message);
         }
     }
 
     /**
-     * @param array $csv
-     * @param bool  $hasHeaders
-     * @param bool  $dryRun
-     *
-     * @return array|null
-     * @throws DatabaseException
+     * @param array $parsedData
+     * @param bool  $dataIsAssociativeArray
+     * @param bool  $flush
      */
-    public function import($csv, $hasHeaders = true, $dryRun = false)
+    protected function importData(array $parsedData, $dataIsAssociativeArray, $flush = true)
     {
-        if (!$this->checkHeadersAreValid($csv)) {
-            return [];
+        $entities = $this->interpreter->interpret($parsedData, $dataIsAssociativeArray);
+
+        $this->log->setImportedEntities($entities);
+
+        if ($flush || $this->log->containErrors()) {
+            $this->detachEntities($entities);
+
+            return;
         }
 
-        $parsedData = $this->parser->parse($csv, $hasHeaders, $this->delimiter);
+        $this->persistEntities($entities);
 
-        return $this->importData($parsedData, $hasHeaders, !$dryRun);
+        try {
+            $this->entityManager->flush();
+        } catch (ORMException $exception) {
+            $message = 'Error when flushing for entity.';
+            $this->handleImporterError(new Error($message, $exception));
+        }
     }
 
     /**
-     * @param string $delimiter
+     * @param array|null $entitiesToPersist
      */
-    public function setDelimiter($delimiter)
+    protected function persistEntities(array $entitiesToPersist = null)
     {
-        $this->delimiter = $delimiter;
+        if (is_null($entitiesToPersist)) {
+            return;
+        }
+
+        foreach ($entitiesToPersist as $entity) {
+            try {
```

Comment:
I wonder how expensive it is to initialize try-catch block. Maybe it's better to:

```
try {
    foreach ...
} catch ...
```